velox/functions/prestosql/geospatial/GeometrySerde.cpp

```diff
@@ -14,28 +14,242 @@
+
+  for (int i = 0; i < partCount; ++i) {
+    startIndexes[i] = input.read<int>();
+  }
+
+  std::vector<int> partLengths(partCount);
+  if (partCount > 1) {
+    partLengths[0] = startIndexes[1];
+    for (int i = 1; i < partCount - 1; ++i) {
+      partLengths[i] = startIndexes[i + 1] - startIndexes[i];
+    }
+  }
+  partLengths[partCount - 1] = pointCount - startIndexes[partCount - 1];
+
+  std::vector<std::unique_ptr<geos::geom::LineString>> lineStrings;
+  lineStrings.reserve(partCount);
+  for (int i = 0; i < partCount; ++i) {
+    lineStrings.push_back(geometryFactory_->createLineString(
+        readCoordinates(input, partLengths[i])));
+  }
+
+  if (multiType) {
+    return geometryFactory_->createMultiLineString(std::move(lineStrings));
+  }
+
+  if (lineStrings.size() != 1) {
+    VELOX_FAIL("Expected a single LineString for non-multiType polyline.");
+  }
+
+  return std::move(lineStrings[0]);
+}
+
+std::unique_ptr<geos::geom::Geometry> GeometryDeserializer::readPolygon(
+    velox::common::InputByteStream& input,
+    bool multiType) {
+  skipEsriType(input);
+  skipEnvelope(input);
+
+  int partCount = input.read<int>();
+  if (partCount == 0) {
+    if (multiType) {
+      return geometryFactory_->createMultiPolygon();
+    }
+    return geometryFactory_->createPolygon();
+  }
+
+  int pointCount = input.read<int>();
+  std::vector<int> startIndexes(partCount);
+  for (int i = 0; i < partCount; i++) {
+    startIndexes[i] = input.read<int>();
+  }
+
+  std::vector<int> partLengths(partCount);
+  if (partCount > 1) {
+    partLengths[0] = startIndexes[1];
+    for (int i = 1; i < partCount - 1; i++) {
+      partLengths[i] = startIndexes[i + 1] - startIndexes[i];
+    }
+  }
+  partLengths[partCount - 1] = pointCount - startIndexes[partCount - 1];
+
+  std::unique_ptr<geos::geom::LinearRing> shell = nullptr;
+  std::vector<std::unique_ptr<geos::geom::LinearRing>> holes;
+  std::vector<std::unique_ptr<geos::geom::Polygon>> polygons;
+
+  for (int i = 0; i < partCount; i++) {
+    auto coordinates = readCoordinates(input, partLengths[i]);
+
+    if (isClockwise(coordinates, 0, coordinates->size())) {
+      // next polygon has started
+      if (shell) {
+        polygons.push_back(geometryFactory_->createPolygon(
+            std::move(shell), std::move(holes)));
+        holes.clear();
+      }
+      shell = geometryFactory_->createLinearRing(std::move(coordinates));
+    } else {
+      holes.push_back(
+          geometryFactory_->createLinearRing(std::move(coordinates)));
+    }
+  }
+  polygons.push_back(
+      geometryFactory_->createPolygon(std::move(shell), std::move(holes)));
+
+  if (multiType) {
+    return geometryFactory_->createMultiPolygon(std::move(polygons));
+  }
+
+  if (polygons.size() != 1) {
+    VELOX_FAIL("Expected exactly one polygon, but found multiple.");
+  }
+  return std::move(polygons[0]);
+}
+
+std::unique_ptr<geos::geom::Geometry> GeometryDeserializer::readEnvelope(
```

Comment:
We should have a function that returns a `geos::geom::Envelope`.  It could be this one, or a separate one.